### ISSUE TITLE
TreeDataProvider documentation is incomplete

### ISSUE THREAD

**Issue by pvh** [Microsoft/vscode#62798]
Right now the TreeDataProvider documentation just reads:
```
## TreeDataProvider<T>
A data provider that provides tree data
```
Obviously this is a little bit thin on the ground, but in implementing support for partial refresh, I discovered that the example code is perhaps misleading. Most of the examples in https://github.com/Microsoft/vscode-extension-samples/tree/master/tree-view-sample don't demonstrate refreshing sub-trees and instead rebuild the entire TreeView on any change.

I suspect, after reading the code of `extHostTreeView.ts` that the arguments to a TreeView's `onDidChangeTreeData` event should be strings ("handles"?) that can be looked up in the this.nodes map but it seems like most of the sample code is recreating objects and not storing them, so those objects can never be refreshed/fetched. (See, for example, [this line in the FTP TreeView example](https://github.com/Microsoft/vscode-extension-samples/blob/master/tree-view-sample/src/fileExplorer.ts#L282).)

At any rate, I suspect an update to the docs and also to the examples is in order and I'm happy to take the time to help with either... if I can know what the right solution is.

**Comment by pvh** [comment]
I think adding the last paragraph from your comment to the documentation would be valuable and illustrative. Several of the examples and external implementations use complex objects which aren't cached anywhere, and the JSON implementation doesn't call attention to it either. 

In the absence of information, I had to dig all the way into the extHostTreeView.ts node refresh code to figure out that these were lookups in a Map object and so comparison was based on reference equality.

I will reiterate that the documentation on TreeDataProvider itself is one sentence which essentially reiterates the name of the class. Of all the vscode APIs I've implemented against it is by far the least documented and had the most surprises.

I'll try and follow-up with a proposed paragraph that says something along the lines of:
 * what a TreeDataProvider is and how it relates to TreeView
 * how the handles work and why it works that way
 * any other suggestions?
